core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -128,15 +154,24 @@ public void clusterChanged(ClusterChangedEvent event) {
                     } else {
                         indexSettings = indexService.getIndexSettings();
                     }
-                    if (indicesService.canDeleteShardContent(shardId, indexSettings)) {
-                        deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                    IndicesService.ShardDeletionCheckResult shardDeletionCheckResult = indicesService.canDeleteShardContent(shardId, indexSettings);
+                    switch (shardDeletionCheckResult) {
+                        case CAN_DELETE:
+                            deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                            break;
+                        case NO_FOLDER_FOUND:
+                        case SHARED_FILE_SYSTEM:
+                            deletionSkipCache.add(shardId);
+                            break;
+                        default:
```

Comment:
Can we add a `case` for `NO_NODE_LOCK` and `STILL_ALLOCATED` that does a `break;` and then have the `default:` case throw an exception, just so in the future if we add another state it doesn't fall through?